contracts/broker/contracts/src/Broker.sol

```diff
@@ -0,0 +1,253 @@
+/*
+
+  Copyright 2019 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.9;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-exchange/contracts/src/interfaces/IExchange.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibAssetDataTransfer.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibFillResults.sol";
+import "@0x/contracts-exchange-libs/contracts/src/LibOrder.sol";
+import "@0x/contracts-utils/contracts/src/LibBytes.sol";
+import "@0x/contracts-utils/contracts/src/LibRichErrors.sol";
+import "@0x/contracts-utils/contracts/src/LibSafeMath.sol";
+import "@0x/contracts-utils/contracts/src/Refundable.sol";
+import "./interfaces/IBroker.sol";
+import "./interfaces/IPropertyValidator.sol";
+import "./libs/LibBrokerRichErrors.sol";
+
+
+// solhint-disable space-after-comma
+contract Broker is
+    Refundable,
+    IBroker
+{
+    bytes[] internal _cachedAssetData;
+    uint256 internal _cacheIndex;
+    address internal _sender;
+    address internal _EXCHANGE; // solhint-disable-line var-name-mixedcase
+
+    using LibSafeMath for uint256;
+    using LibBytes for bytes;
+    using LibAssetDataTransfer for bytes;
+
+    /// @param exchange Address of the 0x Exchange contract.
+    constructor (address exchange)
+        public
+    {
+        _EXCHANGE = exchange;
+    }
+
+    /// @dev A payable fallback function that makes this contract "payable". This is necessary to allow
+    ///      this contract to gracefully handle refunds from the Exchange.
+    function ()
+        external
+        payable
+    {} // solhint-disable-line no-empty-blocks
+
+    /// @dev The Broker implements the ERC1155 transfer function to be compatible with the ERC1155 asset proxy
+    /// @param from Since the Broker serves as the taker of the order, this should equal `address(this)`
+    /// @param to This should be the maker of the order.
+    /// @param amounts Should be an array of just one `uint256`, specifying the amount of the brokered assets to transfer.
+    /// @param data Encodes the validator contract address and any auxiliary data it needs for property validation.
+    function safeBatchTransferFrom(
```

Comment:
OK, so here is my admittedly kind of lame attack vector for this and why we should probably be checking that `msg.sender == ERC1155Proxy`:

- Maker creates one of these orders with `makerFeeAssetData = ERC1155(...)` and `feeRecipient == maker`.
- Taker is like, w/e not my fee not problem so I'll fill this order.
- Taker is also like, my sloppy ass will pass in more ERC721s than the fill/order wants because the broker shouldn't use them anyway.
- Exchange fills taker -> maker, which leaves behind the excess ERC721s "in" the broker contract.
- Exchange then fills maker fee, which is an 1155 so it calls `onERC1155BatchReceived()` on `maker` which, surprise, is a smart contract.
- In `onERC1155BatchReceived()` the maker calls `Broker.safeBatchTransferFrom(address(broker), maker, [], [1], hex"0x")`
- Maker walks away with an extra ERC721.

Reply:
This is actually the scenario that the `if (from != address(this))` protects against. But yeah, I'm going to add a `msg.sender == ERC1155Proxy` check just in case we're missing an edge case